### PR TITLE
Switch to PWA in Firefox extension when SW API is unavailable

### LINKED ISSUE #877: Firefox extension does not switch to the PWA any more since Firefox 103.0
On several devices where I have kiwix-js 3.4.0 Firefox extension installed, it no longer switches to the PWA when trying to use the ServiceWorker mode.
It says "The ServiceWorker API is not available on your device. Falling back to JQuery mode".

I suspect that there has been a change in Firefox behavior, in version 103.0.
In previous version of Firefox, the ServiceWorker API was available in extensions (`'serviceWorker' in navigator` returned `true`), but failed to register. And we trapped this registraion exception to propose a switch to the PWA (when inside a Firefox extension).

In my current version (103.0 64 bits on Ubuntu), `'serviceWorker' in navigator` now returns `false` in the extension, so kiwix-js does not propose to use the PWA, and fallbacks to jQuery mode.

I do not see any mention of this change in https://developer.mozilla.org/en-US/docs/Mozilla/Firefox/Releases/103 so am not 100% sure

### PR DESCRIPTION
**Pull request by mossroy**
This is necessary for Firefox >=103, where the ServiceWorker API has been removed in WebExtensions.
See https://hg.mozilla.org/integration/autoland/rev/3a2907ad88e8

Fixes #877

### CODE REVIEW

**Comment by mossroy** [comment]
Switched this PR to draft as it's not working properly for now

**Comment by mossroy** [comment]
Now tested on Firefox 103.0 on Ubuntu

**Comment by Jaifroid** [comment]
Thanks for this PR. I'll test the UX (by installing as a developer-mode extension).

I wonder if there is another way of preventing users of old Firefox editions (I mean really old ones that don't support Service Worker at all) from being invited to switch to SW mode? Is this a case where testing explicitly for (Firefox < 103 AND SW unavailable) is warranted? Or could we test for a proxy API that will assure us that we are in a recent Firefox?

**Comment by mossroy** [comment]
After fighting to disable auto-update of Firefox, I managed to test with version 52.9.0ESR.
When asking for SW mode, it suggests to use the PWA. If I accept, it opens the PWA, which "The ServiceWorker API is not available on your device" and fallsback to jQuery mode (but stays inside the PWA).
When you close the extension and open it again, it says "The last attempt to launch the PWA appears to have failed. Do you wish to try again?". If you confirm, it switches again to the PWA (same behavior). If you cancel, it stays in local jQuery mode.

So it's not that bad.

If we want to test the browser version, it should only be a "hint", like adding in the systemAlert that it will probably not work because this browser probably does not support ServiceWorkers

**Comment by Jaifroid** [comment]
OK, that's good. I think it's enough for now. We might have to introduce something once we switch to SW mode as default (assuming we will apply that default to the Firefox extension version).

I've tested on:

* Firefox 103 (portable) on Windows 11 in a temporary extension. Works fine.
* Edge for Windows 11 (running from localhost). It starts automatically in SW mode (running from localhost).
* IE11 on Windows 11 (yes, there's a trick to be able to launch IE11 still on Windows 11!!), running from localhost. It rightly tells me that SW mode won't work when I try to switch, and falls back to jQuery mode.
* I haven't tested on old Edge legacy, but it's irrelevant now, and I think the behaviour will just mirror IE11 except that in some desktop versions of Edge legacy SW mode works. In that case it will act the same way as Edge/Chromium.

So I don't think I need to do any more tests. I'll take a look at the code.

File: www/js/app.js [hunk 1]
```diff
@@ -780,76 +780,86 @@ define(['jquery', 'zimArchiveLoader', 'uiUtil', 'settingsStore','abstractFilesys
...
-                                $('#articleContent').contents().remove();
-                                // Create the MessageChannel and send the 'init' message to the ServiceWorker
+            // Since Firefox 103, the ServiceWorker API is not available any more in Webextensions. See https://hg.mozilla.org/integration/autoland/rev/3a2907ad88e8 and https://bugzilla.mozilla.org/show_bug.cgi?id=1593931
+            // Previously, the API was available, but failed to register (which we could trap a fews lines below).
+            // So we now need to suggest a switch to the PWA if we are inside a Firefox Extension and the ServiceWorker API is unavailable.
+            // Even if some older firefox versions do not support ServiceWorkers at all (versions 42, 43, 45ESR, 52ESR, 60ESR and 68ESR, based on https://caniuse.com/serviceworkers). In this case, the PWA will not work either.
+            if (protocol === 'moz-extension:' && !isServiceWorkerAvailable()) {
+                launchMozillaExtensionServiceWorker();
+            }
+            else {
```
Comment (Jaifroid) [inline www/js/app.js:791]: Use same coding style as surrounding code 😉
```suggestion
            } else {
```

**Review by Jaifroid — APPROVED** [review]
Good catch!

Just one small coding-style suggestion below. When you're happy you've done enough testing, feel free to squash/merge.

**Pull request merged**